### PR TITLE
Add bootstrap script

### PR DESCRIPTION
**Pull request by LinusCDE**
Hi,

as suggested by @okeh, I created a script that combines installation, re-enabling, adding toltec and installing a package.

It currently uses [my fork of remarkable_entware](https://github.com/LinusCDE/remarkable_entware/). But in theory it should also work with the original one as it also detects whether toltec is added and adds it automatically.

Since the script my be shared on serveral servers due to the HTTPs implications with the default wget, it has a version number in the comments.

The usage for the one-liner is also in the script. For testing you can use this one-liner:

```
# wget -qO- https://static.cosmos-ink.net/remarkable/bootstrap.sh | sh -s
```

Packages to be installed can be added to the end.

---

@matteodelabre I also saw that you are hosting some other files on toltec as well. You can gladly copy [the wget binary repo](http://static.cosmos-ink.net/remarkable/artifacts) to your repo and use that repo in the file as well.  
My fork for remarkable_install.sh will automatically use that bootstrapped wget. I can also change the url in remarkable_install.sh as well then to have a universal server for all the packages, installers and thirdparty stuff.

### CODE REVIEW

File: bootstrap.sh
Comment (raisjn) [inline bootstrap.sh]: this should be automatic if /opt/bin exists (don't ask user to run the command by hand)

File: bootstrap.sh
Comment (raisjn) [inline bootstrap.sh]: no need to ping remarkable.com, just giving away info to them. maybe hit some other server

File: bootstrap.sh
Comment (raisjn) [inline bootstrap.sh]: is there a common dir, like ~/.config/ or ~/.cache that can be used instead of putting .wget_bin in the toplevel namespace?

File: bootstrap.sh
Comment (raisjn) [inline bootstrap.sh]: toltec stable has almost nothing in it. i feel like we should default to testing or move more stuff to stable. perhaps in the testing -> stable process we have a weekly or biweekly promotion cycle

File: bootstrap.sh
Comment (raisjn) [inline bootstrap.sh]: this should be pointed to toltec-dev/remarkable_entware whenever we can make that repo (imo)

**Review by raisjn — COMMENTED** [review]
we should also have a CI test that uses this script to install a package and verifies that: 1) package was installed, 2) opkg was installed, 3) toltec was enabled

File: bootstrap.sh
Comment (LinusCDE) [reply bootstrap.sh]: Agreed. I'll add that.

File: bootstrap.sh
Comment (LinusCDE) [reply bootstrap.sh]: Is there a good server you'd recommend? I took remarkable.com since it should never be down and everyone would probably trust it. If I'd just taken google.com, some people probably wouldn't have liked that.

Maybe github.com (though there are companies that actually block that domain)?

File: bootstrap.sh
Comment (LinusCDE) [reply bootstrap.sh]: .cache is a good idea. I took this folder, so a user would see it easily. I'll need to change that in my entware_fork then, too.

File: bootstrap.sh
Comment (LinusCDE) [reply bootstrap.sh]: Then I'll remove the explicit mention of Evildo from there and we'll decide that later.

File: bootstrap.sh
Comment (LinusCDE) [reply bootstrap.sh]: Yeah, but the users should have the stable repo. We should really just merge to stable more often. Like, when there are no bug reports for a software in 2 weeks or similar, it's considered stable. I think some distros do it similarly.

File: bootstrap.sh
Comment (Eeems) [reply bootstrap.sh]: Maybe toltec.delab.re or one.one.one.one?

File: bootstrap.sh
Comment (Eeems) [reply bootstrap.sh]: See #20

File: bootstrap.sh
Comment (LinusCDE) [reply bootstrap.sh]: Cloudflare (one.one.one.one) would also be good. I don't want to take toltec.delab.re as this might at some point misreport no internet when there are in fact just maintenance works or similar. The big servers typically are all good for that since they're practicially never down at any given point.

We could also make the point, that I could discard those checks, since in order to run that script, most would download it with wget anyway.

File: bootstrap.sh
Comment (LinusCDE) [reply bootstrap.sh]: Changed to ~/.cache/wget_bin in commit [b712a400](https://github.com/toltec-dev/toltec/pull/38/commits/b712a400c5ffe258e9cc5070adf0c5556af6c323) and [remarkable_entware/52e7c156](https://github.com/LinusCDE/remarkable_entware/commit/52e7c1566d83149894bb4a40b17432bd1c185038).

File: bootstrap.sh
Comment (LinusCDE) [reply bootstrap.sh]: Adjusted in commit [d3b53182](https://github.com/toltec-dev/toltec/pull/38/commits/d3b531825d8b46a1b1b2ab89926e0f39b3e4c61c).

File: bootstrap.sh
Comment (LinusCDE) [reply bootstrap.sh]: I removed those checks. Wouldn't matter if a one-liner tries to download that script in the first place.

**Comment by raisjn** [comment]
I've used this script on three remarkable tablets to great effect! I've put it at https://rmkit.dev/bootstrap and am going to test it from there (not just cosmos-ink) next. 

The only thing that messed  up the script was in one tablet it found /opt/ that was a non opkg /opt and it aborted. I then deleted /opt (because I created it) and re-ran and it worked. When I ran on a tablet with opkg, it correctly recognized /opt and opkg and moved forward. 

My main issue with script is that I had to tell someone to use `testing` repo and that added 2 or 3 extra manual steps.

**Comment by LinusCDE** [comment]
@raisjnn Nice to hear.

> I've put it at https://rmkit.dev/bootstrap and am going to test it from there (not just cosmos-ink) next.

Cool! As it's intended. Also a nice idea to trash the `.sh` extension. Would make any url cleaner.

> My main issue with script is that I had to tell someone to use testing repo and that added 2 or 3 extra manual steps.

I had that yesterday when writing my wall of text which is why the latest commit addresses that. You can add `TOLTEC_BRANCH=testing ` in front of the one-liner to use testing by default.

Though we should really pay attention that this doesn't become the default. It should only ever be used by devs, testers or maintainers. Never appear in a one-liner in some readme.

File: bootstrap.sh
Comment (raisjn) [inline bootstrap.sh]: where is this hosted?

Comment (LinusCDE) [reply bootstrap.sh]: On my Raspberry Pi 4 at home:
- Nginx on Ubuntu Server 20.04
- 30 MBit/s Uplink
- Automatic updating of DNS entries should the ISP have problems (was lately down about 3 times a week for about 3 mins with new ip afterwards)

I can also easily move that to a VPS I have. Or better yet, we just host it under thirdparty on the toltec server.

The file was [built here](https://git.cosmos-ink.net/linus/wget-remarkable-pipeline). You can check out the latest artifact from the CI/CD section there. Should be same hash and commit and build logs are also all there.

Comment (Eeems) [reply bootstrap.sh]: I'd feel much better if it was hosted somewhere that had redundancy.

Comment (matteodelabre) [reply bootstrap.sh]: Is there a reason for not using the [wget package from Entware](http://bin.entware.net/armv7sf-k3.2/wget_1.20.3-4_armv7-3.2.ipk)?

Comment (LinusCDE) [reply bootstrap.sh]: To install entware (mostly atm) from https in the first place.

Comment (matteodelabre) [reply bootstrap.sh]: Here’s an alternative idea: we could run the install script from the context of a host connected to the rM through SSH. That way, we could leverage the up-to-date wget/curl from the host. I believe this would make things simpler. What do you think?

Comment (LinusCDE) [reply bootstrap.sh]: In theory yes, but then we would also need to ensure that user has wget installed. Either downloading it on Windows (and installing on Mac) or requiring docker for just one small installation.  
Considering most end users don't run linux, this would be a hassle for both us and them.

Comment (matteodelabre) [reply bootstrap.sh]: Since the initial `wget -qO- https://static.cosmos-ink.net/remarkable/bootstrap.sh | sh -s` invocation uses the insecure default `wget` from the reMarkable, doesn’t that make all further efforts towards avoiding MITM moot? It seems to me that an attacker can simply MITM the bootstrap.sh file in this case.

Comment (Eeems) [reply bootstrap.sh]: We should include a hash check in that command. Instead of just piping it into sh.

Comment (LinusCDE) [reply bootstrap.sh]: Agreed. That's the dilemma I see currently: Short one-liner vs secure script snippet. We should really go with the latter though.

Comment (LinusCDE) [reply bootstrap.sh]: I hope that the reMarkable team notices this and gives their BusyBox a update (wget is part of that). Then we would use a longer script for about 2 months and could switch to a oneliner afterwards. We could also get rid of any wget bootstrapping.

Comment (Eeems) [reply bootstrap.sh]: The secure snippet wouldn't really be that long though.

**Comment by raisjn** [comment]
are we waiting on anything here or is this ready to go?

**Comment by matteodelabre** [comment]
Overall this looks good, but there are a few nits to fix regarding formatting, variable naming and shellcheck compliance. I’ll review it more thoroughly tomorrow.

File: bootstrap.sh
Comment (matteodelabre) [reply bootstrap.sh]: What about embedding the secure wget binary inside the `bootstrap` script? This way we 1) remove an additional HTTP request, and 2) only need to check the SHA-sum of the downloaded script, instead of both the script and the secure wget binary retrieved by the script.

The embedding could be done automatically by a build script, and the final `bootstrap` script could be automatically published to the Toltec server.

Comment (matteodelabre) [reply bootstrap.sh]: The embedding could be done automatically by a build script, and the final `bootstrap` script could be automatically published to the Toltec server.

Comment (LinusCDE) [reply bootstrap.sh]: > What about embedding the secure wget binary inside the `bootstrap` script? This way we 1) remove an additional HTTP request, and 2) only need to check the SHA-sum of the downloaded script, instead of both the script and the secure wget binary retrieved by the script.
> 
> The embedding could be done automatically by a build script, and the final `bootstrap` script could be automatically published to the Toltec server.

We could probably do that. This would make the script fairly big though. And the fact that the script already has the hash of the weg file builds a chain of trust in that case anyway. If bootstrap.sh has a known hash, that would change if we were to change the hash in that script. It's similar in essence to how git commits work as well, if I'm not mistaken.

EDIT: I also changed the url from above to a github one by now, btw. See the last commit. We could also move that repo into toltec. And have the CI by done by GitHub as well (though I don't know GitHub Actions well enough to to that rn which is why I use the GitLab CI on my server instead).

File: scripts/bootstrap
Comment (matteodelabre) [inline scripts/bootstrap]: Note that this will execute if /home/root/.cache/wget_bin exists and is not a directory. We should probably change the condition to `if [ -f /home/root/.cache/wget_bin/bin ]`.

File: scripts/bootstrap
Comment (matteodelabre) [inline scripts/bootstrap]: This seems dangerous: what if the user happens to have existing data in a folder named like this? I think the script should fail if the directory already exists.

File: scripts/bootstrap
Comment (matteodelabre) [inline scripts/bootstrap]: For consistency, uppercase should be reserved for environment variables, leaving lowercase to normal variables. By the way, all those variables should probably be marked `local`.

Comment (matteodelabre) [reply scripts/bootstrap]: It’s still not clear to me why we can’t manually download the [wget from Entware](http://bin.entware.net/armv7sf-k3.2/wget_1.20.3-4_armv7-3.2.ipk) and extract it to get a secure `wget`, instead of relying on another repository.

File: scripts/bootstrap
Comment (matteodelabre) [inline scripts/bootstrap]: Shouldn’t the `/home/root/.cache/wget_bin` directory get cleaned up when the script ends?

File: scripts/bootstrap
Comment (matteodelabre) [inline scripts/bootstrap]: Couldn’t we merge the `entware_install.sh` script into this script?

File: scripts/bootstrap
Comment (matteodelabre) [inline scripts/bootstrap]: Can we move the `entware_reenable.sh` script into this repository?

File: scripts/bootstrap
Comment (matteodelabre) [inline scripts/bootstrap]: Isn’t this redundant with `entware_install.sh`?

File: scripts/bootstrap
Comment (matteodelabre) [inline scripts/bootstrap]: The `log()` and `throw()` functions in this script could be merged with the `status()`, `section()` and `error()` functions in `scripts/package-lib`.

File: scripts/bootstrap
Comment (matteodelabre) [inline scripts/bootstrap]: Isn’t this script supposed to run on the reMarkable? How can this message be shown if the connection to the reMarkable has been lost?

**Review by matteodelabre — DISMISSED** [review]
Thanks for your work @LinusCDE. Here are some comments.

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: > For consistency, uppercase should be reserved for environment variables, leaving lowercase to normal variables. By the way, all those variables should probably be marked local.

Good find! Should be addressed with commit 35bf74fb.

> It’s still not clear to me why we can’t manually download the wget from Entware and extract it to get a secure wget, instead of relying on another repository.

The "problem" is that entware does a proper job with their packages. They work exclusivly in the /opt directory and don't assume anything about the rest of the system. Even though binaries are dynamically linked, the linker lib itself is always hardcoded in the binary. In case of entwares wget it is on /opt.

Here is an example of me running entwares wget binary without having /opt mounted.
```
reMarkable: ~entwget/ ls
wget      wget-ssl
reMarkable: ~entwget/ ./wget
-sh: ./wget: No such file or directory
reMarkable: ~entwget/ file wget
wget: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /opt/lib/ld-linux.so.3, for GNU/Linux 3.2.0, stripped
reMarkable: ~entwget/ 
```

The `file not found` also confused me first when tinkering with c a while ago. But the interpreter lib `ld-linux.so` is meant with that.

File: scripts/bootstrap
Comment (Eeems) [reply scripts/bootstrap]: It's instructing the user to reconnect after it's done running. Not that the user will lose connection.

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: It seems pretty unlikely that a user would happen to have such a directory to me. Though you're right that this is a concern. Since I added a check for that path to be a file. Doing a rm on a directory is unnecessary and a error for a file is a good idea.

Addressed in commit d8731cd.

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: That is probably a good idea. I didn't do it here, so we may reuse it afterwards if needed.

This commit 6330e6e should address this. It's not perfect but do you find it okay this way (feel free to make changes).

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: Probably. I didn't feel comfortable to make that decision here though. We should talk with @evildo again to maybe get his repo either into [toltec-dev](https://github.com/toltec-dev/) (and him, too) or merge the script outright.

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: See https://github.com/toltec-dev/toltec/pull/38/files/515ea8a7dafcb877ba87a870acc7945f87e1a047#r498383687

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: I don't think so, but I'm not sure. `entware_install.sh` may add `/opt/bin:/opt/sbin` temporarily (not sure on that) and just tells the user how to add the path permanently. The part here does that for the user if opkg was installed automatically or the user forgot it.

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: I'm not entirely sure where see the problem with where it runs. The bootstrap script is always meant to run only on the reMarkable and don't make any assumptions on the users computer os that way.

That said, checking whether the function echo'ed something is still dirty. I also should clean up or reduce or entirely remove `ensure_opkg_available()` and `ensure_opkg_available_from_bashrc()` since they're only run once anyway. Probably a case of YAGNI.

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: Good idea. Though we would then also supply those scripts as well. Do have an idea on how to use them without straight copying them?

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: > It's instructing the user to reconnect after it's done running. Not that the user will lose connection.

Yes. (I didn't saw the comment when I wrote the previous one.)

Is there a way to apply the updated $PATH to the users bash directly. But aside from spawning a child bash with exec I didn't find a good way to do that. So this message was my only solution.

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: See https://github.com/toltec-dev/toltec/pull/38#discussion_r498378708

File: scripts/bootstrap
Comment (Eeems) [reply scripts/bootstrap]: `export PATH=$PATH:...`?

Comment (LinusCDE) [reply scripts/bootstrap]: Only if you source the script. While we could do that in the one liner if someone downloads the script and runs it manually (which we shouldn't discourage imo), it wouldn't work since the export is applied the the child process the script is run in.

Comment (Eeems) [reply scripts/bootstrap]: When running the script source it instead of calling it and you can export variables.

Comment (LinusCDE) [reply scripts/bootstrap]: We could use that for the bootstrap command. But some people may wonder why. Also should people download and execute the script themselves, it wouldn't work.  
Is there a way to detect whether a script was sourced?

Comment (Eeems) [reply scripts/bootstrap]: https://stackoverflow.com/a/2687092

File: scripts/bootstrap
Comment (matteodelabre) [reply scripts/bootstrap]: Good catch, I didn’t think about that! Then I think that it would be best to move the wget build over to this repo, and to publish it on `https://toltec.delab.re`. I’ll do that in a separate PR if it’s OK with you.

File: scripts/bootstrap
Comment (matteodelabre) [reply scripts/bootstrap]: I think a good solution would be to have a generic script which installs Opkg and accepts as arguments a list of repositories to enable. Then it could be used both to install Toltec and to install Entware only. (By the way you probably meant @Evidlo.)

File: scripts/bootstrap
Comment (matteodelabre) [reply scripts/bootstrap]: Okay, so this is related to the thread above about merging with `remarkable_entware`.

File: scripts/bootstrap
Comment (matteodelabre) [reply scripts/bootstrap]: We could have a build step which simply `cat`s the lib and this script, then publishes the result to `toltec.delab.re`.

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: No problem. If you want, I can also move the [wget-remarkable-pipeline](https://github.com/LinusCDE/wget-remarkable-pipeline) over to the toltec-dev domain. Though a custom gitlab actions build would probably be helpful at that point anyway. So we might as well just recreate that repo. What do you think?

Comment (matteodelabre) [reply scripts/bootstrap]: A transfer sounds great, I can add a GitHub workflow to it if you want.

Comment (LinusCDE) [reply scripts/bootstrap]: Is tranferred to toltec-dev. We just to get `buildroot/base:20200814.2228` (from docker hub) to run `ci_build.sh` and zipping the `dist/` folder afterwards.

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: If I'm not mistaken, remarkable_entware basically runs a provided installer from entware (which installs it and automatically adds entware) and fixes some libs as well for the remarkable.  
So we would basically need a combination of this one, entware_install.sh and entwares installer.

File: scripts/bootstrap
Comment (matteodelabre) [reply scripts/bootstrap]: Thanks! I’m not really familiar with buildroot. I wonder if it could be used to simplify the Dockerfiles in toltec-dev/toolchain?

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: Good idea. This would make the repo more complex though. Would be nice if you could add that as I'm not that comfortable with GH Actions, yet.

File: scripts/bootstrap
Comment (LinusCDE) [reply scripts/bootstrap]: > I wonder if it could be used to simplify the Dockerfiles in toltec-dev/toolchain?

Only if you want to build your base image yourself I guess. Buildroot meant to build a Busybox or Systemd bases linux system from scratch along with a bootloader. It uses a similar buildsystem to the Linux kernel and basically downloads and builds every single binary from scratch for many architectures and boards. I use it at work for a RPi-Like board. That way our software will be built into a ready-to-flash sd.img file.

It may be useful for your custom oecore-like toolchain. But I'm not sure whether getting the header files is even a usecase.

File: scripts/bootstrap
Comment (matteodelabre) [reply scripts/bootstrap]: Another drawback is that the resulting file's checksum will change anytime we change either of the `package-lib` or the `bootstrap` script, and therefore we will need to remember to manually update the install command accordingly even if we only update the `package-lib`.

**Comment by matteodelabre** [comment]
I’ve made some changes to the [wget building workflow](https://github.com/toltec-dev/wget). It is now fully standalone (no need to download additional shared libs), which should hopefully make this script a bit simpler. Binaries are automatically pushed to <https://toltec.delab.re/thirdparty/bin> when a release is tagged in toltec-dev/wget. The checksum of the latest version is `8d447c6eb0a39e705f45bea900b12eef07142ea3da0809aca4dd44fe4110cdfd`.

File: scripts/bootstrap
Comment (matteodelabre) [reply scripts/bootstrap]: Leaving this as is for now to move forward.

**Comment by matteodelabre** [comment]
Just changed the script to use Toltec’s wget and to incorporate Entware’s installation script. External dependencies are now reduced to just Entware, which should make auditing the script easier.

It works on my install, but I would appreciate if any of you had time to test it, to find if I introduced regressions. ~~My initial tests seem to show that the error recovery mechanism does not call `error-cleanup` when suitable—something that will need to be fixed before merging.~~ (Fixed.)

Scenarios to test:

- [x] Without any pre-existing Entware install.
- [x] With extra packages to install, passed as arguments to `bootstrap`.
- [x] With the `testing` branch (note that I've lowercased the toltec_branch variable since it's not exported).
- [x] With an existing, enabled remarkable_entware install (it should add Toltec on top).
- [x] With an existing, disabled remarkable_entware install (it should re-enable the /opt mount and add Toltec on top).
- [x] With an existing, enabled Toltec install (it should be a no-op).
- [x] With an existing, disabled Toltec install (it should re-enable the /opt mount).
- [x] With a faulty network that corrupts the initial wget download.
- [x] With a network that becomes unavailable partway through the installation.

Remaining work:

- [x] Automatically publish the `bootstrap` script to `https://toltec.delab.re/bootstrap`, through the `.github/workflow/stable.yml` workflow.
- [x] Update install instructions in README.
- [x] Add license header to the `bootstrap` script.

**Comment by raisjn** [comment]
can you add doc to the top of the script with invocation line? (and example of using wget)

here's what I get with various invocations while testing re-enabling a disabled entware installation:

```
reMarkable: ~/ wget -qO- https://rmkit.dev/bs | sh -s
wget: note: TLS certificate validation not implemented
sh: line 39: syntax error near unexpected token `<'
reMarkable: ~/ wget -qO- https://rmkit.dev/bs | sh -C
wget: note: TLS certificate validation not implemented
sh: line 39: syntax error near unexpected token `<'
reMarkable: ~/ wget -qO- https://rmkit.dev/bs | bash -C
wget: note: TLS certificate validation not implemented
INFO:  Re-enabling existing Entware install
INFO:  Adding /opt/bin to the PATH
INFO:  Use '/home/root/entware-reenable' to re-enable Toltec after a system update
reMarkable: ~/ wget -qO- https://rmkit.dev/bs | bash -C harmony
harmony: harmony: Is a directory
wget: note: TLS certificate validation not implemented
reMarkable: ~/
```

my main feedback here is that i think the `bash -C harmony` line installs harmony, but i don't actually have any confirmation or notices about it. it would be nice for a message to appear indicating which packages were installed. there should also be indication of which repo is being enabled (stable or testing).

**Comment by matteodelabre** [comment]
The errors with `sh` are expected since this is a Bash script.

Note that piping wget output to bash is insecure since it does not support TLS. I couldn’t come up with a simple single-line chain of commands that verifies the script checksum before running it, so for now it’s [documented as three commands to execute](https://github.com/toltec-dev/toltec/pull/38/files#diff-04c6e90faac2675aa89e2176d2eec7d8R15-R17).

You probably meant to use the `-s` Bash flag instead of `-C`. The line you ran tried to read a script from a file called `harmony` which did not exist. If you run `wget -qO- https://rmkit.dev/bs | bash -s harmony`, you’ll see that there is actually feedback when additional packages are installed.

**Comment by matteodelabre** [comment]
> there should also be indication of which repo is being enabled (stable or testing).

Fixed that.

**Comment by raisjn** [comment]
did you add the example invocation to the top of the file? 

> errors expected with sh

unexpected for me, as the script started as an sh script and invocations mention sh.

**Comment by matteodelabre** [comment]
Does this work for you @raisjnn?

~~Ideally we’d also need to make sure that the script’s checksum is always up-to-date in the README.~~ Check added.

File: scripts/bootstrap/bootstrap
Comment (raisjn) [inline scripts/bootstrap/bootstrap:11]: i was hoping it would be a line showing how to use it alongside curl or wget, f.e.: `wget -Qo- <url> | bash -s harmony draft`

**Review by raisjn — APPROVED** [review]
i just want this thing out the door (sorry it's taken so long, linus and matteo)

this script looks good (assuming it works for everyone else) and we can iterate on it after it lands. the bootstrap should ideally still be a one liner in the README, but its good enough for now.

File: scripts/bootstrap/bootstrap
Comment (matteodelabre) [reply scripts/bootstrap/bootstrap:11]: I understand, but I don’t feel comfortable recommending that users run a potentially unsafe chain of commands. And if we make it safe (i.e. include the file’s checksum in the example), then it’s no longer possible to include that line in the file itself.

File: scripts/bootstrap/bootstrap
Comment (Evidlo) [inline scripts/bootstrap/bootstrap:114]: Change this to `<<-` and you can indent the heredoc.

Comment (matteodelabre) [reply scripts/bootstrap/bootstrap:114]: But only if we indent with tabs :(

**Pull request merged**
